cpp/src/arrow/sparse_tensor.cc

```diff
@@ -419,6 +421,152 @@ class SparseTensorConverter<TYPE, SparseCSCIndex>
   inline Status CheckMaximumValue(const uint64_t) const { return Status::OK(); }
 };
 
+// ----------------------------------------------------------------------
+// SparseTensorConverter for SparseCSFIndex
+
+template <typename TYPE>
+class SparseTensorConverter<TYPE, SparseCSFIndex>
+    : private SparseTensorConverterBase<TYPE> {
+ public:
+  using BaseClass = SparseTensorConverterBase<TYPE>;
+  using typename BaseClass::NumericTensorType;
+  using typename BaseClass::value_type;
+
+  SparseTensorConverter(const NumericTensorType& tensor,
+                        const std::shared_ptr<DataType>& index_value_type,
+                        MemoryPool* pool)
+      : BaseClass(tensor, index_value_type, pool) {}
+
+  template <typename IndexValueType>
+  Status Convert() {
+    using c_index_value_type = typename IndexValueType::c_type;
+    RETURN_NOT_OK(CheckMaximumValue(std::numeric_limits<c_index_value_type>::max()));
+
+    const int64_t ndim = tensor_.ndim();
+    std::vector<int64_t> axis_order = internal::ArgSort(tensor_.shape());
+    int64_t nonzero_count = -1;
+    RETURN_NOT_OK(tensor_.CountNonZero(&nonzero_count));
+
+    std::shared_ptr<Buffer> values_buffer;
+    RETURN_NOT_OK(
+        AllocateBuffer(pool_, sizeof(value_type) * nonzero_count, &values_buffer));
+    value_type* values = reinterpret_cast<value_type*>(values_buffer->mutable_data());
+
+    std::vector<int64_t> counts(ndim, 0);
+    std::vector<int64_t> coord(ndim, 0);
+    std::vector<int64_t> previous_coord(ndim, -1);
+    std::vector<TypedBufferBuilder<c_index_value_type>> indptr_buffer_builders(ndim - 1);
+    std::vector<TypedBufferBuilder<c_index_value_type>> indices_buffer_builders(ndim);
+
+    if (ndim <= 1) {
+      return Status::NotImplemented("TODO for ndim <= 1");
+    } else {
+      const std::vector<int64_t>& shape = tensor_.shape();
+      for (int64_t n = tensor_.size(); n > 0; n--) {
+        const value_type x = tensor_.Value(coord);
+
+        if (tensor_.Value(coord) != 0) {
+          bool tree_split = false;
+          *values++ = x;
+
+          for (int64_t i = 0; i < ndim; ++i) {
+            int64_t dimension = axis_order[i];
+            bool change = coord[dimension] != previous_coord[dimension];
+
+            if (tree_split || change) {
+              if (change) tree_split = true;
+
+              if (i < ndim - 1)
+                RETURN_NOT_OK(indptr_buffer_builders[i].Append(
+                    static_cast<c_index_value_type>(counts[dimension + 1])));
+              RETURN_NOT_OK(indices_buffer_builders[i].Append(
+                  static_cast<c_index_value_type>(coord[dimension])));
+              ++counts[dimension];
+            }
+          }
+          previous_coord = coord;
+        }
+
+        // increment index
```

Comment:
Don't we have similar code (incrementing a multi-dimensional index) elsewhere in this file? If reasonably doable, it should be factored out.

Reply:
There is similar code for `COO` sparse tensor but for `CSF` we need to use `axis_order` to correctly increment. We could factor this out but then we need to have `COO` create and pass `axis_order`. I'm not sure that would improve the situation.

Reply:
Related to this, I considered we may need a multi-dimensional index iterator for tensor and sparse tensor before. But I've not added it yet because I've never needed it in the current implementation.

How do you think about extracting index incrementing routines as two functions with the same name: one takes `axis_order`, and the other is specialized for without `axis_order`?

Reply:
I've factored out `IncrementIndex` functions for CSF and COO. Please check if it makes sense. Should we do this for CSR/CSC too?